### DISCUSSION TITLE
ollama-webui is not using GPU and falling back to CPU

### DISCUSSION THREAD

**Discussion by iTestAndroid** [open-webui/open-webui#348] [General]
# Bug Report

## Description

**Bug Summary:**
I have ollama running on background using a model, it's working fine in console, all is good and fast and uses GPU.
I run ollama-webui and I'm not using docker, just did nodejs and uvicorn stuff and it's running on port 8080, it communicated with local ollama I have thats running on 11343 and got the models available. Everything looked fine. As soon as I ask a question, I notice it takes forever because its not using GPU, I check with nvidia-smi in the background, no use. But if I ask the same question in console, I get answers super fast as it uses GPU.

**Steps to Reproduce:**
Just run ollama in background, start ollama-webui locally without docker

**Expected Behavior:**
Reuse existing ollama session and use GPU

**Actual Behavior:**
Ignore GPU all together and fallback to CPU and take forever to answer.

## Environment

- **Operating System:** Ubuntu 22
- **Browser (if applicable):** Chrome

**Comment by tjbck** [comment]
Hi, Thanks for creating this issue! That's seems very strange, as ollama-webui communicates to Ollama via Ollama API routes and as per Ollama's documentation, it should behave exactly the same as using the CLI. Could you try curling to your ollama outside of the webui and try to isolate the problem? Keep us updated, Thanks!

**Comment by iTestAndroid** [comment]
I noticed something else, it seems like on every prompt ollama-webui re-spawns and re-loads  the model? I think probably that's why it seems slow to me.

Is it possible to configure and run llama.cpp once with parameters defined by user, keep the interactive shell up and running and serve requests to it from web UI?

**Comment by tjbck** [comment]
Hmm, I believe this probably is an issue from Ollama-side, not our webui. You might want to create an issue on their repo, I'm sure It will help them a lot! Keep us updated on your journey!

**Reply by justinh-rahb** [reply]
We can now use the `KEEP_ALIVE` parameter to address this.

**Comment by jaifar530** [comment]
I have similar issue posted here #830

**Comment by miguelgargallo** [comment]
Same issue here

**Comment by matth9** [comment]
I am also having the same issue.
I've narrowed it down to using the OLLAMA_HOST environment variable. If this is set, only 1 layer will be offloaded to the gpu, if this is not set and I connect with the default settings then it will offload everything correctly.
I have managed to reproduce this using the ollama cli

**Reply by matth9** [reply]
I spoke too soon, it is only once the model has been loaded by open webui that the issue starts, but once started using the ollama cli also runs slowly until the model is stopped and started again. Even if the model is started in the ollama cli it is always reloaded when used from open webui and then is slow in the cli as well as only 1 layer is moved to the gpu

**Reply by matth9** [reply]
Changing the num_gpu setting from default to 256 appears to have resolved this for me

**Reply by Shreyash113** [reply]
How to change num_gpu setting can u please help me out

**Reply by matth9** [reply]
If you select the controls option in the top right and scroll to the bottom you will see "num_gpu (Ollama)"

**Comment by VladOnMyOwn** [comment]
Same issue here

**Comment by wxzwxz131** [comment]
same problem

**Comment by Alexanderkocha** [comment]
I've noticed this issue on a fresh bundled Ollama and Open Web UI install. Would me running the docker run command with sudo have anything to do with this problem?

**Comment by pflavio** [comment]
I have the same issue, running Ollama in Docker with OpenWebUI. It happens when it isn't used for a bit with sleep/wake cycles in between. Restarting the container fixes it though.

**Comment by kalvinarts** [comment]
I have open-webui running with podman and ollama running as a service on my system.

With `num_gpu` set to **default**, `ollama ps` reports the model is completely loaded on the GPU but all my CPUs get very busy and the GPU is doing nothing:

```
NAME             ID              SIZE     PROCESSOR    UNTIL
qwen3:30b-a3b    2ee832bc15b5    21 GB    100% GPU     4 minutes from now
```

Setting `num_gpu` to `256` the GPU gets busy, but still a couple of my CPUs get to 100% too.

I also had to set `use_mmap` to **disabled** because even with `ollama ps` reporting that the model is only using the GPU, sometimes, the VRAM was not being used.

Running the same model using the ollama CLI, it works as intended, the GPU does all the work and all my CPUs are chilling.

I will try to do some more testing and to see if there are any differences between the API calls from open-webui and ollama CLI.

**Comment by stephen-ritter** [comment]
I was having this issue and was looking through the issues in Ollama, and stumbled onto this:
https://docs.ollama.com/troubleshooting#linux-docker

Basically adding the following to `/etc/docker/daemon.json` seems to have fixed this issue for me:
```
{
    "exec-opts": ["native.cgroupdriver=cgroupfs"]
}
```
It has been running straight for about 2 days without me having to restart the containers.

The associated issue I found this solution in:
https://github.com/ollama/ollama/issues/10257
